### PR TITLE
Fix oneof fields to work properly for basic data types

### PR DESCRIPTION
**Pull request by balusu-murali**
- oneof fields was not working properly for non-message types like int64, bytes, string etc.
- erlang pb libraries in general set the oneof field values to be a tuple with the varname followed by the actual value.
- this is very inconvenient when we write the actual record in erlang code.
- so we made changes to set the oneof field to be the value directly without the variable name.
- changes for that are here: https://github.com/HalloAppInc/enif_protobuf/pull/5 and https://github.com/HalloAppInc/enif_protobuf/pull/2
- this breaks adding basic types as part of oneof because it will be hard to distinguish which field was set.
- so we need to go back to using the variable name for basic types.
- for basic field types: value must be set to {var_name, actual_field_value}
- for message field types: value must be set to the record directly.
- tested this and works well now.

### CODE REVIEW

File: c_src/ep_encoder.c [hunk 1]
```diff
@@ -686,7 +684,28 @@ get_oneof_field(ErlNifEnv *env, ERL_NIF_TERM term, ep_node_t *node, ERL_NIF_TERM
...
-    return bsearch((&(array[0])), node->fields, node->size, sizeof(ep_field_t), get_field_compare_sub_name);
+    // We try and fetch the specific field that was set as part of the oneof field.
+    // oneof field values are set directly to the actual record if the value is a new message.
+    // If the value is one of the basic types - then it is a tuple with the variable name followed by the value.
+    // Here, we first check if the value is another message by checking if the prefix is set to pb_.
+    // If not then we expect the value to be one of the basic types and compare the variable name.
+
+    if(strncmp(buf, "pb_", 3) == 0) {
+        *out = term;
+        _free(buf);
```
Comment (vipin211) [inline c_src/ep_encoder.c]: Is this a bug fix?

File: c_src/ep_encoder.c [hunk 2]
```diff
@@ -686,7 +684,28 @@ get_oneof_field(ErlNifEnv *env, ERL_NIF_TERM term, ep_node_t *node, ERL_NIF_TERM
...
+        return bsearch((&(array[0])), node->fields, node->size, sizeof(ep_field_t), get_field_compare_sub_name);
+    } else {
+        *out = array[1];
+        _free(buf);
+        ep_field_t *result = node->fields;
+        for (int i = 0; i < node->size; ++i) {
+            if (get_field_compare_name((&(array[0])), result) == 0) {
+                return result;
+            }
+            result = node->fields + (i+1)*sizeof(ep_field_t);
```
Comment (vipin211) [inline c_src/ep_encoder.c]: Do you think moving this line before 702 and using i instead of i+1 is better?

File: c_src/ep_encoder.c (hunk 1 above)
Comment (balusu-murali) [reply c_src/ep_encoder.c]: not really a bug - but yeah cleans it up properly.

File: c_src/ep_encoder.c (hunk 2 above)
Comment (balusu-murali) [reply c_src/ep_encoder.c]: sure - can make the change. logic is the same.

File: c_src/ep_encoder.c [hunk 3]
```diff
@@ -686,7 +684,28 @@ get_oneof_field(ErlNifEnv *env, ERL_NIF_TERM term, ep_node_t *node, ERL_NIF_TERM
...
+    if(strncmp(buf, "pb_", 3) == 0) {
+        *out = term;
+        _free(buf);
+        return bsearch((&(array[0])), node->fields, node->size, sizeof(ep_field_t), get_field_compare_sub_name);
+    } else {
+        *out = array[1];
+        _free(buf);
+        ep_field_t *result = node->fields;
+        for (int i = 0; i < node->size; ++i) {
+            result = node->fields + i*sizeof(ep_field_t);
```
Comment (vipin211) [inline c_src/ep_encoder.c]: You can get rid of line 700.
```suggestion
            ep_field_t *result = node->fields + i*sizeof(ep_field_t);
```

File: c_src/ep_encoder.c [hunk 4]
```diff
@@ -675,8 +675,6 @@ get_oneof_field(ErlNifEnv *env, ERL_NIF_TERM term, ep_node_t *node, ERL_NIF_TERM
         return NULL;
     }
 
-    *out = term;
-
     unsigned size = 100;
     char *buf = _alloc(size);
```
Comment (vipin211) [inline c_src/ep_encoder.c]: I am not sure if following is the same. If it is, consider using the following and you will not need _free(buf).
```suggestion
    char buf[100];
```

**Review by vipin211 — APPROVED** [review]

**Review by Thomas-Neill — APPROVED** [review]
LGTM

**Comment by Thomas-Neill** [comment]
the broken test can be fixed by just switching the msg names to use the `pb_` prefix - i have a change i can push if ok

File: c_src/ep_encoder.c [hunk 5]
```diff
@@ -686,7 +684,28 @@ get_oneof_field(ErlNifEnv *env, ERL_NIF_TERM term, ep_node_t *node, ERL_NIF_TERM
...
     }
 
-    return bsearch((&(array[0])), node->fields, node->size, sizeof(ep_field_t), get_field_compare_sub_name);
+    // We try and fetch the specific field that was set as part of the oneof field.
+    // oneof field values are set directly to the actual record if the value is a new message.
+    // If the value is one of the basic types - then it is a tuple with the variable name followed by the value.
+    // Here, we first check if the value is another message by checking if the prefix is set to pb_.
+    // If not then we expect the value to be one of the basic types and compare the variable name.
+
+    if(strncmp(buf, "pb_", 3) == 0) {
```
Comment (loukiano) [inline c_src/ep_encoder.c:692]: can oneof fields be enums? If so I'm a little nervous about casting to `ep_field_t` in the search

Comment (balusu-murali) [reply c_src/ep_encoder.c:692]: it should still work.

**Comment by balusu-murali** [comment]
> LGTM

@Thomas-Neill 
thanks - i also fixed the tests

**Pull request merged**